src/analysis/convex/extreme.lean

```diff
@@ -0,0 +1,210 @@
+- define convex independence and prove lemmas related to extreme points.
+-/
+
+open_locale classical affine
+open set
+
+variables {E : Type*} [add_comm_group E] [vector_space ℝ E] {x : E} {A B C : set E}
+
+/--
+A set B is extreme to a set A if B ⊆ A and all points of B only belong to (intrinsic) interiors of
+segments whose ends are in B.
+-/
+def is_extreme (A B : set E) :
+  Prop :=
+B ⊆ A ∧ ∀ x₁ x₂ ∈ A, ∀ x ∈ B, x ∈ segment x₁ x₂ → x₁ ≠ x → x₂ ≠ x → x₁ ∈ B ∧ x₂ ∈ B
+
+lemma is_extreme.refl :
+  reflexive (is_extreme : set E → set E → Prop) :=
+λ A, ⟨subset.refl _, λ x₁ x₂ hx₁A hx₂A x hxA hx hxx₁ hxx₂, ⟨hx₁A, hx₂A⟩⟩
+
+lemma is_extreme.trans :
+  transitive (is_extreme : set E → set E → Prop) :=
+begin
+  rintro A B C ⟨hBA, hAB⟩ ⟨hCB, hBC⟩,
+  use subset.trans hCB hBA,
+  rintro x₁ x₂ hx₁A hx₂A x hxC hx hxx₁ hxx₂,
+  obtain ⟨hx₁B, hx₂B⟩ := hAB x₁ x₂ hx₁A hx₂A x (hCB hxC) hx hxx₁ hxx₂,
+  exact hBC x₁ x₂ hx₁B hx₂B x hxC hx hxx₁ hxx₂,
+end
+
+lemma is_extreme.antisymm :
+  anti_symmetric (is_extreme : set E → set E → Prop) :=
+λ A B hAB hBA, subset.antisymm hBA.1 hAB.1
+
+instance : is_partial_order (set E) is_extreme :=
+{ refl := is_extreme.refl,
+  trans := is_extreme.trans,
+  antisymm := is_extreme.antisymm }
+
+lemma convex_diff_of_extreme (hA : convex A) (hAB : is_extreme A B) :
+  convex (A \ B) :=
+begin
+  rw convex_iff_segment_subset,
+  rintro x₁ x₂ ⟨hx₁A, hx₁B⟩ ⟨hx₂A, hx₂B⟩ x hx,
+  refine ⟨hA.segment_subset hx₁A hx₂A hx, λ hxB, hx₁B (hAB.2 x₁ x₂ hx₁A hx₂A x hxB hx _ _).1⟩,
+  { rintro rfl,
+    exact hx₁B hxB },
+  { rintro rfl,
+    exact hx₂B hxB }
+end
+
+/--
+A point x is an extreme point of a set A if x belongs to no (intrinsic) interior of a segment with
+ends in A.
+-/
+def set.extreme_points (A : set E) :
+  set E :=
+{x ∈ A | ∀ (x₁ x₂ ∈ A), x ∈ segment x₁ x₂ → x₁ = x ∨ x₂ = x}
+
+lemma extreme_point_def :
+  x ∈ A.extreme_points ↔ x ∈ A ∧ ∀ (x₁ x₂ ∈ A), x ∈ segment x₁ x₂ → x₁ = x ∨ x₂ = x :=
+by refl
+
+/--x is an extreme point to A iff {x} is an extreme set of A.-/
+lemma extreme_point_iff_extreme_singleton :
+  x ∈ A.extreme_points ↔ is_extreme A {x} :=
+begin
+  split,
+  { rintro ⟨hxA, hx⟩,
+    use singleton_subset_iff.2 hxA,
+    rintro x₁ x₂ hx₁A hx₂A y (rfl : y = x) hxs hx₁ hx₂,
+    exfalso,
+    cases hx x₁ x₂ hx₁A hx₂A hxs,
+    exacts [hx₁ h, hx₂ h] },
+  { rintro hx,
+    use singleton_subset_iff.1 hx.1,
+    rintro x₁ x₂ hx₁ hx₂ hxs,
+    by_contra,
+    push_neg at h,
+    exact h.1 (hx.2 x₁ x₂ hx₁ hx₂ x rfl hxs h.1 h.2).1 }
+end
+
+lemma extreme_points_subset :
+  A.extreme_points ⊆ A :=
+λ x hx, hx.1
+
+@[simp]
+lemma extreme_points_empty :
+  (∅ : set E).extreme_points = ∅ :=
+subset_empty_iff.1 extreme_points_subset
+
+@[simp]
+lemma extreme_points_singleton :
+  ({x} : set E).extreme_points = {x} :=
+subset.antisymm extreme_points_subset $ 
+  singleton_subset_iff.2 ⟨mem_singleton x, λ x₁ _ hx₁ _ _, or.inl hx₁⟩
+
+lemma inter_extreme_of_extreme (hAB : is_extreme A B) (hAC : is_extreme A C) :
```

Comment:
Using the name `is_extreme.inter` would let us use dot notation for this lemma: `hAB.inter hAC`, which could be nicer